arduino/quadrature_encoder/quadrature_encoder.ino

```diff
@@ -0,0 +1,363 @@
+
+    uint8_t current_state = (digitalRead(encoder.pin_clock) << 1) | digitalRead(encoder.pin_data);
+
+    //
+    // encoder from(clk | data) and to(clk | data) as a nibble
+    // and use it to look up the direction of that transition
+    //
+    encoder.transition = ((encoder.transition & 0x03) << 2) | current_state;
+    encoder.transitions += TRANSITION_DIRECTION[encoder.transition];
+
+    //
+    // When clock and data pins are high,
+    // then we are at the end of one quadrature
+    // cycle and at the beginning of the next.
+    // If we had 4 valid transitions, count the tick,
+    // otherwise there was an invalid transition (a bounce), 
+    // so we do not count the tick.
+    //
+    if (0x03 == current_state) {
+      if (4 == encoder.transitions) {
+        encoder.ticks += 1; // we went 1 cycle clockwise
+      } else if (-4 == encoder.transitions) {
+        encoder.ticks -= 1;
+      }
+      encoder.transitions = 0;
+    }
+  #endif
+}
+
+
+void pollEncoders() {
+  for(uint8_t i = 0; i < ENCODER_COUNT; i += 1) {
+    pollEncoder(encoders[i]);
+  }
+}
+
+//
+// reset tick counters to zero
+//
+void resetEncoders() {
+  // turn off interrupts so we can
+  // write mutable shared state atomically
+  #ifdef USE_ENCODER_INTERRUPTS
+     noInterrupts();
+  #endif
+
+  for(uint8_t i = 0; i < ENCODER_COUNT; i += 1) {
+    encoders[i].ticks = 0;
+  }
+  
+  // turn back on interrupts
+  #ifdef USE_ENCODER_INTERRUPTS
+    interrupts();
+  #endif
+}
+
+void readEncoders(long ticks[ENCODER_COUNT]) {
+  // turn off interrupts so we can
+  // read mutable shared state atomically
+  #ifdef USE_ENCODER_INTERRUPTS
+     noInterrupts();
+  #endif
+
+  for(uint8_t i = 0; i < ENCODER_COUNT; i += 1) {
+    ticks[i] = encoders[i].ticks;
+  }
+
+  // turn back on interrupts
+  #ifdef USE_ENCODER_INTERRUPTS
+    interrupts();
+  #endif
+}
+
+
+//
+// write ticks to serial port
+// as a string tuple with ticks and tickMs separate by a comma
+//
+//   "{ticks},{ticksMs}"
+//
+void writeTicks(unsigned long ticksMs) {
+  long ticks[ENCODER_COUNT];
+  readEncoders(ticks);
+
+  Serial.print(ticks[0]);
+  Serial.print(',');
+  Serial.print(ticksMs);
+  for(uint8_t i = 1; i < ENCODER_COUNT; i += 1) {
+    Serial.print(";");
+    Serial.print(ticks[i]);
+    Serial.print(',');
+    Serial.print(ticksMs);
+  }
+  Serial.println("");
+}
+
+
+void setup() {
+  // set all encoder pins to inputs
+  for(uint8_t i = 0; i < ENCODER_COUNT; i += 1) {
+    pinMode(encoders[i].pin_clock, INPUT);
+    pinMode(encoders[i].pin_data, INPUT);
+    #ifdef USE_ENCODER_INTERRUPTS
+      //
+      // if the pin is interrupt capable, then use interrupts
+      //
+      if (NOT_AN_INTERRUPT != digitalPinToInterrupt(encoders[i].pin_clock)) {
+        attachInterrupt(digitalPinToInterrupt(encoders[i].pin_clock), _isr_routines[i], FALLING);
+      }
+    #endif
+  }
+
+  Serial.begin(115200);
+}
+
+
+void loop() {
```

Comment:
No, I don't want to try to share code between those sketches.  Sharing code is not an Arduino strong point.  Beyond that, connecting the two sketches in that way I think would produce more issues down the road due to coupling that it would save.  If we had a lot of these then it could be worth it, but we only have the two.

Reply:
Yeah the overhead of factoring out is maybe not justifying the effort. The real use will only come if someone improves, extends, or compresses the code in one of these files, then we'll miss the same operation on the other, if we don't bear in mind that these are actually meant to be doing the same thing.